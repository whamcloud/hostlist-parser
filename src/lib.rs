// Copyright (c) 2019 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

mod structures;

use crate::structures::{flatten_ranges, Part, RangeOutput};
use combine::{
    attempt, between,
    char::{alpha_num, digit, spaces},
    choice,
    combinator::ignore,
    eof,
    error::{ParseError, StreamError},
    many1, not_followed_by, optional,
    parser::repeat::repeat_until,
    parser::EasyParser,
    sep_by1,
    stream::{Stream, StreamErrorFor},
    token, Parser,
};
use itertools::Itertools;
use std::collections::BTreeSet;

fn comma<I>() -> impl Parser<I, Output = char>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    token(',')
}

fn open_bracket<I>() -> impl Parser<I, Output = char>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    token('[')
}

fn close_bracket<I>() -> impl Parser<I, Output = char>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    token(']')
}

fn dash<I>() -> impl Parser<I, Output = char>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    token('-')
}

fn optional_spaces<I>() -> impl Parser<I, Output = Option<()>>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    optional(spaces())
}

fn host_elements<I>() -> impl Parser<I, Output = String>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    many1(alpha_num().or(dash()).or(token('.')))
}

fn digits<I>() -> impl Parser<I, Output = String>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    many1(digit())
}

fn parsed_digits<I>() -> impl Parser<I, Output = u64>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    digits().and_then(|x| x.parse::<u64>().map_err(StreamErrorFor::<I>::other))
}

fn leading_zeros<I>() -> impl Parser<I, Output = (usize, u64)>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    digits().and_then(|x| {
        let mut digits = x.chars().take_while(|x| x == &'0').count();

        if x.len() == digits {
            digits -= 1;
        }

        x.parse::<u64>()
            .map(|num| (digits, num))
            .map_err(StreamErrorFor::<I>::other)
    })
}

fn range_digits<I>() -> impl Parser<I, Output = RangeOutput>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    attempt((
        leading_zeros(),
        optional_spaces().with(dash()),
        optional_spaces().with(leading_zeros()),
    ))
    .and_then(|((start_zeros, start), _, (end_zeros, end))| {
        let mut xs = vec![start, end];
        xs.sort();

        let (start, end, start_zeros, end_zeros) = if start > end {
            (end, start, end_zeros, start_zeros)
        } else {
            (start, end, start_zeros, end_zeros)
        };

        if end_zeros > start_zeros {
            Err(StreamErrorFor::<I>::unexpected_static_message(
                "larger end padding",
            ))
        } else {
            Ok(RangeOutput::Range(start_zeros, start, end))
        }
    })
}

fn disjoint_digits<I>() -> impl Parser<I, Output = RangeOutput>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    let not_name = not_followed_by(
        optional_spaces()
            .with(digits())
            .skip(optional_spaces())
            .skip(dash())
            .map(|_| ""),
    );

    sep_by1(
        optional_spaces()
            .with(parsed_digits())
            .skip(optional_spaces()),
        attempt(comma().skip(not_name)),
    )
    .map(RangeOutput::Disjoint)
}

fn range<I>() -> impl Parser<I, Output = Vec<RangeOutput>>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    between(
        open_bracket(),
        close_bracket(),
        sep_by1(range_digits().or(disjoint_digits()), comma()),
    )
}

fn hostlist<I>() -> impl Parser<I, Output = Vec<Part>>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    repeat_until(
        choice([
            range().map(Part::Range).left(),
            optional_spaces()
                .with(host_elements())
                .map(Part::String)
                .right(),
        ]),
        attempt(optional_spaces().skip(ignore(comma()).or(eof()))),
    )
    .and_then(|xs: Vec<_>| {
        if xs.is_empty() {
            Err(StreamErrorFor::<I>::unexpected_static_message(
                "no host found",
            ))
        } else {
            Ok(xs)
        }
    })
}

fn hostlists<I>() -> impl Parser<I, Output = Vec<Vec<Part>>>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    sep_by1(hostlist(), optional_spaces().with(comma()))
}

pub fn parse(
    input: &str,
) -> Result<BTreeSet<String>, combine::stream::easy::Errors<char, &str, usize>> {
    let (hosts, _) = hostlists()
        .easy_parse(input)
        .map_err(|err| err.map_position(|p| p.translate_position(input)))?;

    let mut xs = BTreeSet::new();

    for parts in hosts {
        let x_prod: Vec<_> = parts
            .iter()
            .filter_map(Part::get_ranges)
            .map(|xs| flatten_ranges(xs))
            .multi_cartesian_product()
            .collect();

        // No ranges means no interpolation
        if x_prod.is_empty() {
            if parts.is_empty() {}

            let mut s = String::new();

            for p in parts.clone() {
                if let Part::String(x) = p {
                    s.push_str(&x)
                }
            }

            xs.insert(s);
        } else {
            for ys in x_prod {
                let mut it = ys.iter();

                let mut s = String::new();

                for p in parts.clone() {
                    match p {
                        Part::String(x) => s.push_str(&x),
                        Part::Range(_) => s.push_str(it.next().unwrap()),
                    }
                }

                xs.insert(s);
            }
        }
    }

    Ok(xs)
}

#[cfg(test)]
mod tests {
    use super::*;
    use combine::parser::EasyParser;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_leading_zeros() {
        assert_debug_snapshot!(leading_zeros().easy_parse("001"));
        assert_debug_snapshot!(leading_zeros().easy_parse("0001"));
        assert_debug_snapshot!(leading_zeros().easy_parse("01"));
        assert_debug_snapshot!(leading_zeros().easy_parse("00"));
        assert_debug_snapshot!(leading_zeros().easy_parse("0"));
    }

    #[test]
    fn test_range_digits() {
        assert_debug_snapshot!(range_digits().easy_parse("001-003"));
        assert_debug_snapshot!(range_digits().easy_parse("001 -  003"));
        assert_debug_snapshot!(range_digits().easy_parse("1-100"));
        assert_debug_snapshot!(range_digits().easy_parse("100-0"));
    }

    #[test]
    fn test_disjoint_digits() {
        assert_debug_snapshot!(disjoint_digits().easy_parse("1,2,3,4,5]"));
        assert_debug_snapshot!(disjoint_digits().easy_parse("1,2,3-5"));
    }

    #[test]
    fn test_range() {
        assert_debug_snapshot!(range().easy_parse("[1,2,3,4,5]"));
        assert_debug_snapshot!(range().easy_parse("[1,2,3-5]"));
        assert_debug_snapshot!(range().easy_parse("[1,2,3-5,6,7,8-10]"));
        assert_debug_snapshot!(range().easy_parse("[01-10]"));
    }

    #[test]
    fn test_hostlist() {
        assert_debug_snapshot!(hostlist().easy_parse("oss1.local"));
        assert_debug_snapshot!(hostlist().easy_parse("oss[1,2].local"));
        assert_debug_snapshot!(hostlist().easy_parse(
            "hostname[2,6,7].iml.com,hostname[10,11-12,2-3,5].iml.com,hostname[15-17].iml.com"
        ));
    }

    #[test]
    fn test_hostlists() {
        assert_debug_snapshot!(hostlists().easy_parse("oss1.local"));
        assert_debug_snapshot!(hostlists().easy_parse("oss[1,2].local"));
        assert_debug_snapshot!(hostlists().easy_parse(
            "hostname[2,6,7].iml.com,hostname[10,11-12,2-3,5].iml.com,hostname[15-17].iml.com"
        ));
        assert_debug_snapshot!(hostlists().easy_parse(
            "hostname[2,6,7].iml.com, hostname[10,11-12,2-3,5].iml.com, hostname[15-17].iml.com"
        ));
    }

    #[test]
    fn test_parse() {
        assert_debug_snapshot!(parse("oss[1,2].local"));

        assert_debug_snapshot!(parse("oss1.local"));

        assert_debug_snapshot!(parse("hostname[10,11-12,002-003,5].iml.com"));

        assert_debug_snapshot!(parse(
            "hostname[2,6,7].iml.com,hostname[10,11-12,2-3,5].iml.com,hostname[15-17].iml.com"
        ));

        assert_debug_snapshot!(parse(
            "hostname[2,6,7].iml.com,  hostname[10,11-12,2-3,5].iml.com, hostname[15-17].iml.com"
        ));

        assert_debug_snapshot!("single item without ranges", parse("hostname1.iml.com"));

        assert_debug_snapshot!(
            "two items without ranges",
            parse("hostname1.iml.com, hostname2.iml.com")
        );

        assert_debug_snapshot!(
            "single item with single range",
            parse("hostname[6].iml.com")
        );

        assert_debug_snapshot!(
            "single item with single range and nothing after the range",
            parse("hostname[6]")
        );

        assert_debug_snapshot!(
            "single item with single digit prefixed range",
            parse("hostname[09-11]")
        );
        assert_debug_snapshot!(
            "single item with double digit prefixed range",
            parse("hostname[009-011]")
        );

        assert_debug_snapshot!("single item in reverse order", parse("hostname[7-5]"));

        assert_debug_snapshot!(
            "multiple items combined into regular and reverse order",
            parse("hostname[7-5], hostname[8,9], hostname[3,2,1]")
        );

        assert_debug_snapshot!("long range with prefix", parse("hostname[001-999]"));

        assert_debug_snapshot!(
            "single item with two ranges",
            parse("hostname[6,7]-[9-11].iml.com")
        );

        assert_debug_snapshot!(
            "single item with range containing mixture of comma and dash",
            parse("hostname[7,9-11].iml.com")
        );

        assert_debug_snapshot!(
            "Single range per hostname with dup",
            parse(
                "hostname[2,6,7].iml.com,hostname[10,11-12,2-4,5].iml.com, hostname[15-17].iml.com"
            )
        );

        assert_debug_snapshot!("Multiple ranges per hostname in which the difference is 1", parse("hostname[1,2-3].iml[2,3].com,hostname[3,4,5].iml[2,3].com,hostname[5-6,7].iml[2,3].com"));

        assert_debug_snapshot!(
            "Multiple ranges per hostname in which the difference is 1 two formats",
            parse("hostname[1,2-3].iml[2,3].com,hostname[1,2,3].iml[2,4].com")
        );

        assert_debug_snapshot!(
            "Multiple ranges per hostname in which the difference is gt 1",
            parse("hostname[1,2-3].iml[2,3].com,hostname[4,5].iml[3,4].com")
        );

        assert_debug_snapshot!(
            "no prefix to prefix should throw an error",
            parse("hostname[9-0011]").unwrap_err()
        );

        assert_debug_snapshot!(
            "Overlapping ranges",
            parse("hostname[1,2-3].iml[2,3].com,hostname[3,4,5].iml[3,4].com")
        );

        assert_debug_snapshot!(
            "Duplicate without using a range",
            parse("hostname4.iml.com,hostname4.iml.com")
        );

        assert_debug_snapshot!(
            "Single item with single range and additional characters after range",
            parse("hostname[0-3]-eth0.iml.com")
        );

        assert_debug_snapshot!(
            "Single item with three ranges separated by character",
            parse("hostname[1,2]-[3-4]-[5,6].iml.com")
        );

        assert_debug_snapshot!(
            "Single item with two ranges and no separation between the ranges",
            parse("hostname[1,2][3,4].iml.com")
        );

        assert_debug_snapshot!(
            "Single item with prefix range starting at 0",
            parse("test[000-002].localdomain")
        );

        assert_debug_snapshot!(
            "Combined Ranges",
            parse(
                "hostname[2,6,7].iml.com,hostname[10,11-12,2-3,5].iml.com, hostname[15-17].iml.com"
            )
        );

        assert_debug_snapshot!(
            "Padding with a single and double digit number",
            parse("hostname[06-10]")
        );

        assert_debug_snapshot!(
            "Invalid character in range (snowman)",
            parse("test[00☃-002].localdomain")
        );

        assert_debug_snapshot!("Empty expression", parse(""));

        assert_debug_snapshot!(
            "No separation between comma's",
            parse("hostname[1,,2].iml.com")
        );

        assert_debug_snapshot!(
            "No separation between dashes",
            parse("hostname[1--2].iml.com")
        );

        assert_debug_snapshot!(
            "No separation between dash and comma",
            parse("hostname[1-,2].iml.com")
        );

        assert_debug_snapshot!(
            "No separation between comma and dash",
            parse("hostname[1,-2].iml.com")
        );

        assert_debug_snapshot!("Missing closing brace", parse("hostname[1"));

        assert_debug_snapshot!("Ending an expression with a comma", parse("hostname[1],"));

        assert_debug_snapshot!(
            "Beginning and ending prefixes that don't match with two single digit numbers",
            parse("hostname[01-009]")
        );

        assert_debug_snapshot!(
            "Having a closing brace before an opening brace",
            parse("hostname]00[asdf")
        );
    }
}
