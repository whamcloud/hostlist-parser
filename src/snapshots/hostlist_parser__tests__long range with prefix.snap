---
source: src/lib.rs
assertion_line: 361
expression: "parse(\"hostname[001-999]\")"

---
Ok(
    [
        "hostname001",
        "hostname002",
        "hostname003",
        "hostname004",
        "hostname005",
        "hostname006",
        "hostname007",
        "hostname008",
        "hostname009",
        "hostname010",
        "hostname011",
        "hostname012",
        "hostname013",
        "hostname014",
        "hostname015",
        "hostname016",
        "hostname017",
        "hostname018",
        "hostname019",
        "hostname020",
        "hostname021",
        "hostname022",
        "hostname023",
        "hostname024",
        "hostname025",
        "hostname026",
        "hostname027",
        "hostname028",
        "hostname029",
        "hostname030",
        "hostname031",
        "hostname032",
        "hostname033",
        "hostname034",
        "hostname035",
        "hostname036",
        "hostname037",
        "hostname038",
        "hostname039",
        "hostname040",
        "hostname041",
        "hostname042",
        "hostname043",
        "hostname044",
        "hostname045",
        "hostname046",
        "hostname047",
        "hostname048",
        "hostname049",
        "hostname050",
        "hostname051",
        "hostname052",
        "hostname053",
        "hostname054",
        "hostname055",
        "hostname056",
        "hostname057",
        "hostname058",
        "hostname059",
        "hostname060",
        "hostname061",
        "hostname062",
        "hostname063",
        "hostname064",
        "hostname065",
        "hostname066",
        "hostname067",
        "hostname068",
        "hostname069",
        "hostname070",
        "hostname071",
        "hostname072",
        "hostname073",
        "hostname074",
        "hostname075",
        "hostname076",
        "hostname077",
        "hostname078",
        "hostname079",
        "hostname080",
        "hostname081",
        "hostname082",
        "hostname083",
        "hostname084",
        "hostname085",
        "hostname086",
        "hostname087",
        "hostname088",
        "hostname089",
        "hostname090",
        "hostname091",
        "hostname092",
        "hostname093",
        "hostname094",
        "hostname095",
        "hostname096",
        "hostname097",
        "hostname098",
        "hostname099",
        "hostname100",
        "hostname101",
        "hostname102",
        "hostname103",
        "hostname104",
        "hostname105",
        "hostname106",
        "hostname107",
        "hostname108",
        "hostname109",
        "hostname110",
        "hostname111",
        "hostname112",
        "hostname113",
        "hostname114",
        "hostname115",
        "hostname116",
        "hostname117",
        "hostname118",
        "hostname119",
        "hostname120",
        "hostname121",
        "hostname122",
        "hostname123",
        "hostname124",
        "hostname125",
        "hostname126",
        "hostname127",
        "hostname128",
        "hostname129",
        "hostname130",
        "hostname131",
        "hostname132",
        "hostname133",
        "hostname134",
        "hostname135",
        "hostname136",
        "hostname137",
        "hostname138",
        "hostname139",
        "hostname140",
        "hostname141",
        "hostname142",
        "hostname143",
        "hostname144",
        "hostname145",
        "hostname146",
        "hostname147",
        "hostname148",
        "hostname149",
        "hostname150",
        "hostname151",
        "hostname152",
        "hostname153",
        "hostname154",
        "hostname155",
        "hostname156",
        "hostname157",
        "hostname158",
        "hostname159",
        "hostname160",
        "hostname161",
        "hostname162",
        "hostname163",
        "hostname164",
        "hostname165",
        "hostname166",
        "hostname167",
        "hostname168",
        "hostname169",
        "hostname170",
        "hostname171",
        "hostname172",
        "hostname173",
        "hostname174",
        "hostname175",
        "hostname176",
        "hostname177",
        "hostname178",
        "hostname179",
        "hostname180",
        "hostname181",
        "hostname182",
        "hostname183",
        "hostname184",
        "hostname185",
        "hostname186",
        "hostname187",
        "hostname188",
        "hostname189",
        "hostname190",
        "hostname191",
        "hostname192",
        "hostname193",
        "hostname194",
        "hostname195",
        "hostname196",
        "hostname197",
        "hostname198",
        "hostname199",
        "hostname200",
        "hostname201",
        "hostname202",
        "hostname203",
        "hostname204",
        "hostname205",
        "hostname206",
        "hostname207",
        "hostname208",
        "hostname209",
        "hostname210",
        "hostname211",
        "hostname212",
        "hostname213",
        "hostname214",
        "hostname215",
        "hostname216",
        "hostname217",
        "hostname218",
        "hostname219",
        "hostname220",
        "hostname221",
        "hostname222",
        "hostname223",
        "hostname224",
        "hostname225",
        "hostname226",
        "hostname227",
        "hostname228",
        "hostname229",
        "hostname230",
        "hostname231",
        "hostname232",
        "hostname233",
        "hostname234",
        "hostname235",
        "hostname236",
        "hostname237",
        "hostname238",
        "hostname239",
        "hostname240",
        "hostname241",
        "hostname242",
        "hostname243",
        "hostname244",
        "hostname245",
        "hostname246",
        "hostname247",
        "hostname248",
        "hostname249",
        "hostname250",
        "hostname251",
        "hostname252",
        "hostname253",
        "hostname254",
        "hostname255",
        "hostname256",
        "hostname257",
        "hostname258",
        "hostname259",
        "hostname260",
        "hostname261",
        "hostname262",
        "hostname263",
        "hostname264",
        "hostname265",
        "hostname266",
        "hostname267",
        "hostname268",
        "hostname269",
        "hostname270",
        "hostname271",
        "hostname272",
        "hostname273",
        "hostname274",
        "hostname275",
        "hostname276",
        "hostname277",
        "hostname278",
        "hostname279",
        "hostname280",
        "hostname281",
        "hostname282",
        "hostname283",
        "hostname284",
        "hostname285",
        "hostname286",
        "hostname287",
        "hostname288",
        "hostname289",
        "hostname290",
        "hostname291",
        "hostname292",
        "hostname293",
        "hostname294",
        "hostname295",
        "hostname296",
        "hostname297",
        "hostname298",
        "hostname299",
        "hostname300",
        "hostname301",
        "hostname302",
        "hostname303",
        "hostname304",
        "hostname305",
        "hostname306",
        "hostname307",
        "hostname308",
        "hostname309",
        "hostname310",
        "hostname311",
        "hostname312",
        "hostname313",
        "hostname314",
        "hostname315",
        "hostname316",
        "hostname317",
        "hostname318",
        "hostname319",
        "hostname320",
        "hostname321",
        "hostname322",
        "hostname323",
        "hostname324",
        "hostname325",
        "hostname326",
        "hostname327",
        "hostname328",
        "hostname329",
        "hostname330",
        "hostname331",
        "hostname332",
        "hostname333",
        "hostname334",
        "hostname335",
        "hostname336",
        "hostname337",
        "hostname338",
        "hostname339",
        "hostname340",
        "hostname341",
        "hostname342",
        "hostname343",
        "hostname344",
        "hostname345",
        "hostname346",
        "hostname347",
        "hostname348",
        "hostname349",
        "hostname350",
        "hostname351",
        "hostname352",
        "hostname353",
        "hostname354",
        "hostname355",
        "hostname356",
        "hostname357",
        "hostname358",
        "hostname359",
        "hostname360",
        "hostname361",
        "hostname362",
        "hostname363",
        "hostname364",
        "hostname365",
        "hostname366",
        "hostname367",
        "hostname368",
        "hostname369",
        "hostname370",
        "hostname371",
        "hostname372",
        "hostname373",
        "hostname374",
        "hostname375",
        "hostname376",
        "hostname377",
        "hostname378",
        "hostname379",
        "hostname380",
        "hostname381",
        "hostname382",
        "hostname383",
        "hostname384",
        "hostname385",
        "hostname386",
        "hostname387",
        "hostname388",
        "hostname389",
        "hostname390",
        "hostname391",
        "hostname392",
        "hostname393",
        "hostname394",
        "hostname395",
        "hostname396",
        "hostname397",
        "hostname398",
        "hostname399",
        "hostname400",
        "hostname401",
        "hostname402",
        "hostname403",
        "hostname404",
        "hostname405",
        "hostname406",
        "hostname407",
        "hostname408",
        "hostname409",
        "hostname410",
        "hostname411",
        "hostname412",
        "hostname413",
        "hostname414",
        "hostname415",
        "hostname416",
        "hostname417",
        "hostname418",
        "hostname419",
        "hostname420",
        "hostname421",
        "hostname422",
        "hostname423",
        "hostname424",
        "hostname425",
        "hostname426",
        "hostname427",
        "hostname428",
        "hostname429",
        "hostname430",
        "hostname431",
        "hostname432",
        "hostname433",
        "hostname434",
        "hostname435",
        "hostname436",
        "hostname437",
        "hostname438",
        "hostname439",
        "hostname440",
        "hostname441",
        "hostname442",
        "hostname443",
        "hostname444",
        "hostname445",
        "hostname446",
        "hostname447",
        "hostname448",
        "hostname449",
        "hostname450",
        "hostname451",
        "hostname452",
        "hostname453",
        "hostname454",
        "hostname455",
        "hostname456",
        "hostname457",
        "hostname458",
        "hostname459",
        "hostname460",
        "hostname461",
        "hostname462",
        "hostname463",
        "hostname464",
        "hostname465",
        "hostname466",
        "hostname467",
        "hostname468",
        "hostname469",
        "hostname470",
        "hostname471",
        "hostname472",
        "hostname473",
        "hostname474",
        "hostname475",
        "hostname476",
        "hostname477",
        "hostname478",
        "hostname479",
        "hostname480",
        "hostname481",
        "hostname482",
        "hostname483",
        "hostname484",
        "hostname485",
        "hostname486",
        "hostname487",
        "hostname488",
        "hostname489",
        "hostname490",
        "hostname491",
        "hostname492",
        "hostname493",
        "hostname494",
        "hostname495",
        "hostname496",
        "hostname497",
        "hostname498",
        "hostname499",
        "hostname500",
        "hostname501",
        "hostname502",
        "hostname503",
        "hostname504",
        "hostname505",
        "hostname506",
        "hostname507",
        "hostname508",
        "hostname509",
        "hostname510",
        "hostname511",
        "hostname512",
        "hostname513",
        "hostname514",
        "hostname515",
        "hostname516",
        "hostname517",
        "hostname518",
        "hostname519",
        "hostname520",
        "hostname521",
        "hostname522",
        "hostname523",
        "hostname524",
        "hostname525",
        "hostname526",
        "hostname527",
        "hostname528",
        "hostname529",
        "hostname530",
        "hostname531",
        "hostname532",
        "hostname533",
        "hostname534",
        "hostname535",
        "hostname536",
        "hostname537",
        "hostname538",
        "hostname539",
        "hostname540",
        "hostname541",
        "hostname542",
        "hostname543",
        "hostname544",
        "hostname545",
        "hostname546",
        "hostname547",
        "hostname548",
        "hostname549",
        "hostname550",
        "hostname551",
        "hostname552",
        "hostname553",
        "hostname554",
        "hostname555",
        "hostname556",
        "hostname557",
        "hostname558",
        "hostname559",
        "hostname560",
        "hostname561",
        "hostname562",
        "hostname563",
        "hostname564",
        "hostname565",
        "hostname566",
        "hostname567",
        "hostname568",
        "hostname569",
        "hostname570",
        "hostname571",
        "hostname572",
        "hostname573",
        "hostname574",
        "hostname575",
        "hostname576",
        "hostname577",
        "hostname578",
        "hostname579",
        "hostname580",
        "hostname581",
        "hostname582",
        "hostname583",
        "hostname584",
        "hostname585",
        "hostname586",
        "hostname587",
        "hostname588",
        "hostname589",
        "hostname590",
        "hostname591",
        "hostname592",
        "hostname593",
        "hostname594",
        "hostname595",
        "hostname596",
        "hostname597",
        "hostname598",
        "hostname599",
        "hostname600",
        "hostname601",
        "hostname602",
        "hostname603",
        "hostname604",
        "hostname605",
        "hostname606",
        "hostname607",
        "hostname608",
        "hostname609",
        "hostname610",
        "hostname611",
        "hostname612",
        "hostname613",
        "hostname614",
        "hostname615",
        "hostname616",
        "hostname617",
        "hostname618",
        "hostname619",
        "hostname620",
        "hostname621",
        "hostname622",
        "hostname623",
        "hostname624",
        "hostname625",
        "hostname626",
        "hostname627",
        "hostname628",
        "hostname629",
        "hostname630",
        "hostname631",
        "hostname632",
        "hostname633",
        "hostname634",
        "hostname635",
        "hostname636",
        "hostname637",
        "hostname638",
        "hostname639",
        "hostname640",
        "hostname641",
        "hostname642",
        "hostname643",
        "hostname644",
        "hostname645",
        "hostname646",
        "hostname647",
        "hostname648",
        "hostname649",
        "hostname650",
        "hostname651",
        "hostname652",
        "hostname653",
        "hostname654",
        "hostname655",
        "hostname656",
        "hostname657",
        "hostname658",
        "hostname659",
        "hostname660",
        "hostname661",
        "hostname662",
        "hostname663",
        "hostname664",
        "hostname665",
        "hostname666",
        "hostname667",
        "hostname668",
        "hostname669",
        "hostname670",
        "hostname671",
        "hostname672",
        "hostname673",
        "hostname674",
        "hostname675",
        "hostname676",
        "hostname677",
        "hostname678",
        "hostname679",
        "hostname680",
        "hostname681",
        "hostname682",
        "hostname683",
        "hostname684",
        "hostname685",
        "hostname686",
        "hostname687",
        "hostname688",
        "hostname689",
        "hostname690",
        "hostname691",
        "hostname692",
        "hostname693",
        "hostname694",
        "hostname695",
        "hostname696",
        "hostname697",
        "hostname698",
        "hostname699",
        "hostname700",
        "hostname701",
        "hostname702",
        "hostname703",
        "hostname704",
        "hostname705",
        "hostname706",
        "hostname707",
        "hostname708",
        "hostname709",
        "hostname710",
        "hostname711",
        "hostname712",
        "hostname713",
        "hostname714",
        "hostname715",
        "hostname716",
        "hostname717",
        "hostname718",
        "hostname719",
        "hostname720",
        "hostname721",
        "hostname722",
        "hostname723",
        "hostname724",
        "hostname725",
        "hostname726",
        "hostname727",
        "hostname728",
        "hostname729",
        "hostname730",
        "hostname731",
        "hostname732",
        "hostname733",
        "hostname734",
        "hostname735",
        "hostname736",
        "hostname737",
        "hostname738",
        "hostname739",
        "hostname740",
        "hostname741",
        "hostname742",
        "hostname743",
        "hostname744",
        "hostname745",
        "hostname746",
        "hostname747",
        "hostname748",
        "hostname749",
        "hostname750",
        "hostname751",
        "hostname752",
        "hostname753",
        "hostname754",
        "hostname755",
        "hostname756",
        "hostname757",
        "hostname758",
        "hostname759",
        "hostname760",
        "hostname761",
        "hostname762",
        "hostname763",
        "hostname764",
        "hostname765",
        "hostname766",
        "hostname767",
        "hostname768",
        "hostname769",
        "hostname770",
        "hostname771",
        "hostname772",
        "hostname773",
        "hostname774",
        "hostname775",
        "hostname776",
        "hostname777",
        "hostname778",
        "hostname779",
        "hostname780",
        "hostname781",
        "hostname782",
        "hostname783",
        "hostname784",
        "hostname785",
        "hostname786",
        "hostname787",
        "hostname788",
        "hostname789",
        "hostname790",
        "hostname791",
        "hostname792",
        "hostname793",
        "hostname794",
        "hostname795",
        "hostname796",
        "hostname797",
        "hostname798",
        "hostname799",
        "hostname800",
        "hostname801",
        "hostname802",
        "hostname803",
        "hostname804",
        "hostname805",
        "hostname806",
        "hostname807",
        "hostname808",
        "hostname809",
        "hostname810",
        "hostname811",
        "hostname812",
        "hostname813",
        "hostname814",
        "hostname815",
        "hostname816",
        "hostname817",
        "hostname818",
        "hostname819",
        "hostname820",
        "hostname821",
        "hostname822",
        "hostname823",
        "hostname824",
        "hostname825",
        "hostname826",
        "hostname827",
        "hostname828",
        "hostname829",
        "hostname830",
        "hostname831",
        "hostname832",
        "hostname833",
        "hostname834",
        "hostname835",
        "hostname836",
        "hostname837",
        "hostname838",
        "hostname839",
        "hostname840",
        "hostname841",
        "hostname842",
        "hostname843",
        "hostname844",
        "hostname845",
        "hostname846",
        "hostname847",
        "hostname848",
        "hostname849",
        "hostname850",
        "hostname851",
        "hostname852",
        "hostname853",
        "hostname854",
        "hostname855",
        "hostname856",
        "hostname857",
        "hostname858",
        "hostname859",
        "hostname860",
        "hostname861",
        "hostname862",
        "hostname863",
        "hostname864",
        "hostname865",
        "hostname866",
        "hostname867",
        "hostname868",
        "hostname869",
        "hostname870",
        "hostname871",
        "hostname872",
        "hostname873",
        "hostname874",
        "hostname875",
        "hostname876",
        "hostname877",
        "hostname878",
        "hostname879",
        "hostname880",
        "hostname881",
        "hostname882",
        "hostname883",
        "hostname884",
        "hostname885",
        "hostname886",
        "hostname887",
        "hostname888",
        "hostname889",
        "hostname890",
        "hostname891",
        "hostname892",
        "hostname893",
        "hostname894",
        "hostname895",
        "hostname896",
        "hostname897",
        "hostname898",
        "hostname899",
        "hostname900",
        "hostname901",
        "hostname902",
        "hostname903",
        "hostname904",
        "hostname905",
        "hostname906",
        "hostname907",
        "hostname908",
        "hostname909",
        "hostname910",
        "hostname911",
        "hostname912",
        "hostname913",
        "hostname914",
        "hostname915",
        "hostname916",
        "hostname917",
        "hostname918",
        "hostname919",
        "hostname920",
        "hostname921",
        "hostname922",
        "hostname923",
        "hostname924",
        "hostname925",
        "hostname926",
        "hostname927",
        "hostname928",
        "hostname929",
        "hostname930",
        "hostname931",
        "hostname932",
        "hostname933",
        "hostname934",
        "hostname935",
        "hostname936",
        "hostname937",
        "hostname938",
        "hostname939",
        "hostname940",
        "hostname941",
        "hostname942",
        "hostname943",
        "hostname944",
        "hostname945",
        "hostname946",
        "hostname947",
        "hostname948",
        "hostname949",
        "hostname950",
        "hostname951",
        "hostname952",
        "hostname953",
        "hostname954",
        "hostname955",
        "hostname956",
        "hostname957",
        "hostname958",
        "hostname959",
        "hostname960",
        "hostname961",
        "hostname962",
        "hostname963",
        "hostname964",
        "hostname965",
        "hostname966",
        "hostname967",
        "hostname968",
        "hostname969",
        "hostname970",
        "hostname971",
        "hostname972",
        "hostname973",
        "hostname974",
        "hostname975",
        "hostname976",
        "hostname977",
        "hostname978",
        "hostname979",
        "hostname980",
        "hostname981",
        "hostname982",
        "hostname983",
        "hostname984",
        "hostname985",
        "hostname986",
        "hostname987",
        "hostname988",
        "hostname989",
        "hostname990",
        "hostname991",
        "hostname992",
        "hostname993",
        "hostname994",
        "hostname995",
        "hostname996",
        "hostname997",
        "hostname998",
        "hostname999",
    ],
)
